File: Domain/ChemicalEquation.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AngryElectron.Domain
{
    public class ChemicalEquation : IEnumerable<IChemical>
    {
        public EquationSide Reactants = new EquationSide();
        public EquationSide Products = new EquationSide();

        public ChemicalEquation()
        {
        }

        public ChemicalEquation(string inputString)
        {
            ChemicalEquation myEquation = Parser.Parse(inputString);
            Reactants = myEquation.Reactants;
            Products = myEquation.Products;
        }
        
        public int MoleculeCount { get { return Reactants.Count + Products.Count; } }
        public bool IsBalanced { get { return (Math.Round(Reactants.Mass, 4) == Math.Round(Products.Mass, 4)); } }
        public Element[] Elements { get { return Reactants.Elements.Union(Products.Elements).ToArray(); } }

        public void Add(IChemical chemical, Side side)
        {
            if (side == Side.LeftSide)
                Reactants.Add(chemical);
            if (side == Side.RightSide)
                Products.Add(chemical);
        }

        public IEnumerator<IChemical> GetEnumerator()
        {
            var unionList = Products.Union(Reactants).ToList();
            foreach (var chemical in unionList)
            {
                if (chemical == null)
                    break;
                else
                    yield return chemical;
            }
        }

        public override string ToString()
        {
            StringBuilder sb = new StringBuilder();
            sb.Append(Reactants.ToString());
            sb.Append(" -> ");
            sb.Append(Products.ToString());
            return sb.ToString();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return this.GetEnumerator();
        }

        public string ToStringWithoutCoefficients()
        {
            return Reactants.ToStringWithoutCoefficients() + " -> " + Products.ToStringWithoutCoefficients();
        }

        public string ToHtml()
        {
            StringBuilder sb = new StringBuilder();
            sb.Append(Reactants.ToHTML());
            sb.Append(" -> ");
            sb.Append(Products.ToHTML());
            return sb.ToString();
        }

        public int GetDeepElementCount(Element element)
        {
            return Products.GetDeepElementCount(element) + Reactants.GetDeepElementCount(element);
        }
    }
}

File: Presentation/Web/Scripts/AngryElectron/ChemicalEquation.js
var ChemicalEquation = function(reactants, products){
    return {
        reactants: reactants,
        products: products,
        toString: function () {
            return this.reactants.toString() + " --> " + this.products.toString();
        },
        toHtml: function() {
            return this.reactants.toHtml() + " &rarr; " + this.products.toHtml();
        }
    }
}

var ChemicalExpression = function (chemicals) {
    expressionChemicals;
    return {
        chemicals: chemicals,
        toString: function () {
            var expression = "";
            var isAdditional = false;
            for (var chemical in chemicals) {
                if (isAdditional) { expression += " + "; }
                expression += chemical.toString();
                isAdditional = true;
            }
            return expression;
        },
        toHtml: function() {
            var expression = "";
            var isAdditional = false;
            for (var i ; i < chemicals.count ; i++) {
                if (isAdditional) { expression += " + "; }
                expression += chemical[i].toString();
                isAdditional = true;
            }
            return expression;
        }
    }
}

var Chemical = function (statement) {
    
}